common/persistence/history_task_queue_manager.go

```diff
@@ -0,0 +1,185 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/persistence/serialization"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+const (
+	// clusterNamesHashSuffixLength is the number of characters to use from the hash of the cluster names when forming
+	// the queue name. This is used to avoid name collisions when a cluster name contains the separator character.
+	clusterNamesHashSuffixLength = 8
+)
+
+var (
+	ErrSerializeTaskToEnqueue = errors.New("failed to serialize history task for task queue")
+	// ErrDeserializeRawHistoryTask is returned when the raw task cannot be deserialized from the task queue. This error
+	// is returned when this whole top-level proto cannot be deserialized.
+	//  Raw Task (a proto): <-- when this cannot be deserialized
+	//	- ShardID
+	//	- Blob (a serialized task)
+	ErrDeserializeRawHistoryTask = errors.New("failed to deserialize raw history task from task queue")
+	// ErrDeserializeHistoryTask is returned when the history task cannot be deserialized from the task queue. This
+	// error is returned when the blob inside the raw task cannot be deserialized.
+	//  Raw Task (a proto):
+	//	- ShardID
+	//	- Blob (a serialized task) <-- when this cannot be deserialized
+	ErrDeserializeHistoryTask       = errors.New("failed to deserialize history task blob")
```

Comment:
should those errors use `SerializationError`, `DeserializationError` or `UnknownEncodingTypeError` type?

Reply:
Fixed to use the first two.